### PR TITLE
[NOID] Add reactive hydration on apoc.export.cypher.query

### PR DESCRIPTION
**Pull request by BennuFire**
Based on https://github.com/neo4j-contrib/neo4j-apoc-procedures/pull/3639

Created from an internal branch now.

### CODE REVIEW

File: core/src/main/java/org/neo4j/cypher/export/CypherDatabaseSubGraph.java [hunk 1]
```diff
@@ -0,0 +1,223 @@
...
+import java.util.Optional;
+import java.util.stream.Collectors;
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
+
+import static apoc.export.cypher.formatter.CypherFormatterUtils.cypherNode;
+import static apoc.util.Util.quote;
+import static org.neo4j.internal.kernel.api.TokenRead.ANY_LABEL;
+
+public class CypherDatabaseSubGraph implements SubGraph
```
Comment (conker84) [inline core/src/main/java/org/neo4j/cypher/export/CypherDatabaseSubGraph.java:40]: this class, unfortunately, has some problems. Lemme explain, `CypherResultSubGraph` provides information only for the graph entities involved in the subgraph, unlikely this one. Please check the following comments

File: core/src/main/java/org/neo4j/cypher/export/CypherDatabaseSubGraph.java [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+        return this.relationshipTypes;
+    }
+
+    @Override
+    public Iterable<Label> getAllLabelsInUse() {
+        return this.labels;
+    }
+
+    @Override
+    public long countsForRelationship(Label start, RelationshipType type, Label end) {
```
Comment (conker84) [inline core/src/main/java/org/neo4j/cypher/export/CypherDatabaseSubGraph.java:162]: this method returns the true count even if one of the input elements is not in the provided Cypher query.

File: core/src/main/java/org/neo4j/cypher/export/CypherDatabaseSubGraph.java [hunk 3]
```diff
@@ -0,0 +1,223 @@
...
+        String startNode = cypherNode(start);
+        String endNode = cypherNode(end);
+        String relationship = String.format("[r:%s]", quote(type.name()));
+        return transaction.execute(String.format("MATCH %s-%s->%s RETURN count(r) AS count", startNode, relationship, endNode))
+                .<Long>columnAs("count")
+                .next();
+    }
+
+    @Override
+    public long countsForNode(Label label) {
```
Comment (conker84) [inline core/src/main/java/org/neo4j/cypher/export/CypherDatabaseSubGraph.java:188]: same as above, it returns the real count even if the label is not in the cypher query

File: core/src/main/java/org/neo4j/cypher/export/CypherDatabaseSubGraph.java [hunk 4]
```diff
@@ -0,0 +1,223 @@
...
+                .next();
+    }
+
+    private boolean labelNotExists(Label label, int labelId) {
+        // `label == null` means that we want to count all the labels
+        // otherwise if we don't found a not-null label, we want the count to return to zero
+        return label != null && labelId == ANY_LABEL;
+    }
+
+    private Integer getLabelId(Label start, TokenRead tokenRead) {
```
Comment (conker84) [inline core/src/main/java/org/neo4j/cypher/export/CypherDatabaseSubGraph.java:212]: same as above

File: core/src/main/java/org/neo4j/cypher/export/CypherDatabaseSubGraph.java [hunk 5]
```diff
@@ -0,0 +1,223 @@
...
+
+    @Override
+    public Iterable<IndexDefinition> getIndexes(Label label) {
+        return transaction.schema().getIndexes(label);
+    }
+
+    @Override
+    public Iterable<IndexDefinition> getIndexes(RelationshipType type) {
+        return transaction.schema().getIndexes(type);
+    }
```
Comment (conker84) [inline core/src/main/java/org/neo4j/cypher/export/CypherDatabaseSubGraph.java:149]: these methods return the real information even if the label or the relationship type is not in the subgraph.

**Review by conker84 — COMMENTED** [review]
@BennuFire thank you so much please check my comments.
In the end, maybe is worth investigating what's the cause of the OOM. If you could provide any info about how to reproduce the problem, we're glad to support you in the resolution.

**Comment by BennuFire** [comment]
> @BennuFire thank you so much please check my comments. In the end, maybe is worth investigating what's the cause of the OOM. If you could provide any info about how to reproduce the problem, we're glad to support you in the resolution.

Hi @conker84 thanks for your comment. If you wanna investigate on the OOM, I have created a random database of (around) 300MB with `call apoc.generate.ba(1000000, 8, null, null)`. 

Then I have evaluated the behavior of 

`CALL apoc.export.cypher.query(
    "//nono 
    MATCH ()-[r]->() RETURN *",
    null, {
        stream : true
    }
) YIELD cypherStatements
RETURN count(*)` 

Around 1GB of heap, the procedure stops OOM'ing but give no statements as a result, and starts progressively giving results (count increase) the more heap you give to the db. So it silently fails for unaware users.

**Comment by conker84** [comment]
> > @BennuFire thank you so much please check my comments. In the end, maybe is worth investigating what's the cause of the OOM. If you could provide any info about how to reproduce the problem, we're glad to support you in the resolution.
> 
> Hi @conker84 thanks for your comment. If you wanna investigate on the OOM, I have created a random database of (around) 300MB with `call apoc.generate.ba(1000000, 8, null, null)`.
> 
> Then I have evaluated the behavior of
> 
> `CALL apoc.export.cypher.query( "//nono MATCH ()-[r]->() RETURN *", null, { stream : true } ) YIELD cypherStatements RETURN count(*)`
> 
> Around 1GB of heap, the procedure stops OOM'ing but give no statements as a result, and starts progressively giving results (count increase) the more heap you give to the db. So it silently fails on unaware users.

@BennuFire I tried with 

APOC 4.4.0.19, Neo4j 4.4.23 with default configuration

in particular these values:

```
dbms.memory.heap.initial_size=512m
dbms.memory.heap.max_size=1G
dbms.memory.pagecache.size=512m
```


with your same procedure calls and I wasn't able to reproduce the problem. 

Do you have any more info that you can share in order to reproduce it?

**Comment by conker84** [comment]
I'm closing this for lack of feedback. Feel free to reopen to start again the discussion.
Thank you so much!

**Pull request closed without merging**